### ISSUE TITLE
Equivalent of SQL CASE

### ISSUE THREAD

**Issue by edwindj** [hadley/dplyr#631]
In data cleaning data often contains values that one would like to impute/replace.

For example: 

In a survey for smokers, one would not ask a person his or her cigarette consumption when he/she has indicated not to be a smoker. So technically the number of cigarretes is not measured, but it can be deduced to be zero.

```S
smoke <- read.csv(textConnection(
"name,is_smoker,cigarrettes,cigares
 John,TRUE,15, 0
 Adam,FALSE,NA,NA
 Phil,TRUE,0,1
"
))

#standard R way
smoke$cigarrettes[!smoke$is_smoker & is.na(smoke$cigarrettes)] <- 0
```
How does one do this elegantly in dplyr?

My 2 cents:
```S
#dplyr option 1: split the data set, fix one part and combine the two
smoker_true <- smoke %>% filter(!is_smoker, is.na(smoke$cigarettes))
# complement
smoker_false <- smoke %>% filter(is_smoker | !is.na(smoke$cigarettes))

smoke2 <- rbind( smoker_true %>% mutate(cigarrettes=0)
               , smoker_false
)
```
which I don't find elegant dplyr code.

So try a little harder
```S
# dplyr option 2: replace the whole variable and do the filtering in the assignment
smoke2 <- 
  smoke %>%
  mutate(cigarrettes=ifelse(!is_smoker & is.na(cigarrettes), 0, cigarrettes))
```
It works, but this is not efficient/workable for the database backends and I find it unelegant to
do the filtering in the assignment.

I suggest a new method:
`mutate_where` that allows for changing values of _existing_ variables of a subset of a `tbl`.
So something like
```S
smoke2 <-
   smoke %>% mutate_where(.where=!is_smoker&is.na(cigarrettes), cigarettes=0)
```  
Or another option (in analogy with `group` and `grouped_<tbl>`) to change `filter` or introduce `filtered_<tbl>`/add_filter: setting a filter lets one work on a subset of the data, but one can `remove_filter`.

```S
smoke2 <-
   smoke %>% 
   add_filter(is_smoker, is.na(cigarrettes) %>%
   mutate(cigarrettes=0) %>%
   remove_filter()
```

Any better ideas or suggestions?

**Comment by mattwg** [comment]
I currently solve this problem as follows:
```
df <- data.frame(duration=c(seq(1,28),NA))

df %>%
    mutate(my_duration=
                       ifelse(is.na(duration),0,
                       ifelse(duration==6,1,
                       ifelse(duration==7,1,
                       ifelse(duration==8,1,
                       ifelse(duration==13,2,
                       ifelse(duration==14,2,
                       ifelse(duration==15,2,
                       ifelse(duration==20,3,
                       ifelse(duration==21,3,
                       ifelse(duration==22,3,
                       ifelse(duration==27,4,
                       ifelse(duration==28,4,
                       ifelse(duration==29,4,
                       ifelse(is.na(duration),-1,0))))))))))))))
           )
```
Way too many parentheses :-) but I think it is readable - and by dealing with NA first you are OK.

Matt

**Comment by hadley** [comment]
@mattwg `?cut`

**Comment by mattwg** [comment]
Umm..not very pretty though:

```df <- data.frame(duration=c(seq(1,28),NA))

df %>%
    mutate(my_duration=cut(duration,breaks=c(-1,4,7,11,14,18,21,25,28,40),labels=c(-1,1,-2,2,-3,3,-4,4,-5))) %>%
    filter(!is.na(my_duration), as.numeric(as.character(my_duration))>0)
```

**Comment by matthieugomez** [comment]
I like the merge syntax but this does a deep copy of `df`.

**Comment by rgknight** [comment]
Lay person here... my suggestion is to have the solution accept more than one variable, e.g., 

recode(var1:var4,  -99 = NA)

as this would let us benefit from the : method of selecting variables. Unless this is already possible?

**Comment by nirski** [comment]
@mattwg I like the "piped" version of `ifelse` more, it's much cleaner:

```{R}
mutate(df, my_duration = 0 %>%
    ifelse(duration == 6, 1, .) %>%
    ifelse(duration == 13, 2, .) %>%
    ...
)
```

**Comment by bergsmat** [comment]
Late to the discussion, but I have this problem, too (subset assignment).  For instance, how to rewrite `within` (below) using, e.g. `mutate`?
```
class(Theoph$Time) <- 'myTime'
Theoph$later <- Theoph$Time + 0.01
#Theoph <- within(Theoph, Time[Time < 2] <- later[Time < 2])
Theoph %<>% mutate(Time = ifelse(Time < 2, Time, later ))
class(Theoph$Time) # numeric
```
`ifelse` is unsatisfying because it drops special classes (as illustrated).  Also, I avoid `ifelse` because it silently returns NA when `test` is NA, which is technically correct but usually not what I want.

**Comment by matthieugomez** [comment]
I'd like a modifier of mutate that allows to apply a function to a subset. For instance to compute the cumulative sum only for these observations with id == 1, return NA otherwise, one needs to do:

df %>% mutate(temp = ifelse(id==1, v, 0)
df %>% mutate(temp = cumsum(temp))
df %>% mutate(cumsum = ifelse(id==1, temp, NA)

It's a different user case than recode (in particular it's about one subset rather than a partition) so it probably deserves a different name.

**Comment by src1371** [comment]
I've been using dplyr a lot recently and very happy with it.  Just a question about combining select() and filter();  it would be great to be able to select columns based on a filter eg  data_forPCA <- data_forPCA[, colSums(is.na(data_forPCA)) < 100]  is there a dplyr way of doing this?

**Comment by holstius** [comment]
For what it's worth: here's a `patch` verb that I'm starting to use in my own work:

  https://gist.github.com/holstius/cbc4ec40057fbc2d9f4b

In my workflows, I like to separate "patching" from "recoding" or "labeling". Patching is the real-world consequence of deadlines + busy upstream providers — it's not supposed to be there, but it is.

**Comment by gergness** [comment]
I've been using @juba's code above, but found a bug when trying to set values to NA (they are treated as the default case). In case anyone else is using it, here's my work around:

```
recode <- function(df, ..., match = c("first", "last")) {
  match <- match.arg(match)
  
  cases <- lapply(list(...), as.case)
  if (identical(match, "last"))
    cases <- rev(cases)
  
  tmpval <- lapply(cases, function(case) {
    eval(case$val, df, case$env)
  })
  
  tmpexpr <- lapply(cases, function(case) {
    if (isTRUE(case$expr)) return(TRUE)
    eval(case$expr, df, case$env)
  })
  
  first_true <- apply(data.frame(tmpexpr), 1, function(y) min(which(y)))
  out_val <- as.matrix(data.frame(tmpval, stringsAsFactors = FALSE))
  
  vapply(seq_along(first_true),
         function(x) out_val[x, first_true[x]],
         out_val[1])
}
```

**Comment by cronacronis** [comment]
Why not to separate selection of what we want to change with chaning? E.g. `mark %>% mutate`, whether `mark` just stores attributes which rows should be affected by `mutate`?  Main difference between `mark` and `filter` that, filter *removes* rows from output, whether mark just passes what to mutate further keeping full output.

**Comment by hadley** [comment]
With the new `if_else` and `coalesce` functions, you could now write the original motivating example as

```R
mutate(smoke, cigarettes = if_else(coalesce(is_smoker, FALSE), cigarettes, 0L))
```

And note that this discussions is really about a sql CASE equivalent for R, not a vectorised switch. The discussion for vectorised switch is now taking place in #1710. 

I've also deleted a number of comments in this thread that are no longer pertinent.

**Comment by hadley** [comment]
And now you can do:

```R
mutate(smoke, cigarettes = if_else(is_smoker, cigarettes, 0L, 0L))
```

**Comment by hadley** [comment]
I'm pretty happy with this implementation, but I'd love to get everyone's feedback on the name and

**Comment by lionel-** [comment]
These new functions are awesome!


##### case_when()

It could be named `match_cases()` or `match_patterns()`.

I find the tilde a bit hard to parse visually with a long list of cases. Maybe allow the `:=` operator? The only thing to do would be to export `` `:=` <- `~` ``.

```{r}
match_cases(
  x %% 35 == 0 := "fizz buzz",
  x %% 5 == 0 := "fizz",
  x %% 7 == 0 := "buzz",
  TRUE := as.character(x)
)
```


##### recode()

Would it be too unsafe to coerce the default value automatically?

```{r}
recode(x, 5, 4, 3, default = x)
#> Error: `default` has type 'integer' not 'double'

recode(x, "a", "b", "c", default = x)
#> Error: `default` has type 'integer' not 'character'

recode(x, "a", "b", "c", default = NA)
#> Error: `default` has type 'logical' not 'character'
```

against the more verbose:

```{r}
recode(x, 5, 4, 3, default = as.double(x))
recode(x, "a", "b", "c", default = as.character(x))
recode(x, "a", "b", "c", default = NA_character_)
```

**Comment by hadley** [comment]
I think I prefer `~` for now - you can always add more spaces around it to make it more obvious.

I'm also in an explicit mood at the moment, so I'd prefer to require the default to be correctly typed. Also the logic starts to get a bit hairy if there are no arguments (just the default).

**Comment by lionel-** [comment]
I think `recode()` will be very hard to use with numeric data. It would be great to be able to do:

```r
recode(x, `1` = 5, `2` = 4)
```

This would ideally work with both integers and integer-like doubles (as often integer data is stored in a double).

**Comment by gergness** [comment]
Agreed, looks great! 

What about using the formula notation from `case_when` for `recode`? This would help with lionel's point about numeric data, and also could prevent name conflicts for `x`, `default` and `missing`.

```r
x <- c("x", "y", "z")
recode(x, x = "a", y = "b", z = "c")
#>Error: All replacements must be named
```

**Comment by kevinushey** [comment]
I also don't like using named arguments for `recode()` as the names are forced to be / become symbols, whereas `~` or `:=` preserve the LHS type. And `recode(x, `1` = 1)` is pretty ugly.

**Comment by hadley** [comment]
@kevinushey yeah, but I don't see that being a particularly common use case. And I think using `~` means that we'll need to do some work to make `case_when()` work with the hybrid evaluator.

**Comment by krlmlr** [comment]
In 4920c8b5f, I need to use `.$` to access columns when using `case_when()` in `mutate()` -- is this intended?

```
> mtcars %>% tbl_df %>%  mutate(size = case_when(cyl > 6 ~ "big", TRUE ~ "small"))
Error: object 'cyl' not found
> mtcars %>% tbl_df %>%  transmute(size = case_when(cyl > 6 ~ "big", TRUE ~ "small"))
Error: object 'cyl' not found
> mtcars %>% tbl_df %>%  transmute(size = case_when(.$cyl > 6 ~ "big", TRUE ~ "small"))
Source: local data frame [32 x 1]

    size
   (chr)
1  small
2  small
3  small
4  small
5    big
6  small
7    big
8  small
9  small
10 small
..   ...
```

**Comment by krlmlr** [comment]
Re formula notation for recode(): It's also a question of API consistency.

**Comment by hadley** [comment]
@krlmlr see #1719

I think the point of `if_else`, `recode` and `case_when` is not to have exactly equivalent interfaces but to solve a series of increasingly more complicated/general situations that each require progressively more syntax.

**Comment by krlmlr** [comment]
I agree to "more syntax", but to me this doesn't necessary imply that each syntax should be different.

A different implementation of case_when(), that is also consistent with `if_else()` and doesn't suffer from #1719, would be

```
case_when <- function(condition, result, ...) {
  if (missing(condition)) NULL
  else
    if_else(condition, result, case_when(...))
}
```

(replace the if_else() call with its body).

**Comment by sfirke** [comment]
re: @krlmlr at https://github.com/hadley/dplyr/issues/631#issuecomment-197224743, is `.$` indeed the intended syntax?  I ask because I was going to suggest adding an example of `case_when` with `mutate` to the documentation.  Use with `mutate` will be common and some users might need a pointer toward `.$`. since in their current nested `ifelse` calls they don't need it.

Example:
```
mtcars %>% mutate(
  efficiency = case_when(
    .$mpg < 22 ~ "low",
    .$mpg <= 26  ~ "moderate",
    TRUE ~ "high"
  )
)
```

**Comment by hadley** [comment]
@sfirke no - the goals is for `case_when()` to work as is inside `mutate()`.

**Comment by bergsmat** [comment]
Loving case_when, but encountering results I don't fully understand (dplyr 3074c).
```
x <- expand.grid(y=5,z = c(T,F))
x <-  group_by(x,z)
mutate(          x, y = case_when(   z ~ n()         )) # (no error)
mutate(          x, y = case_when(   z ~ n(),!z ~ 0L )) # (no error)
mutate( ungroup(x), y = case_when(   z ~ n(),!z ~ 0L )) # object 'z' not found
mutate(          x, y = case_when(   z ~ 1L          )) # object 'z' not found
mutate(          x, y = case_when( .$z ~ 1L          )) # object 'z' not found
```

**Comment by hadley** [comment]
Yes, it doesn't yet work inside mutate.

**Comment by Diego-MX** [comment]
Hey there dplayers. 
I've seen this is a longstanding issue that is slowing cooking very nicely.  I've seen several posts that address this issue in different scenarios, and can only amaze myself about the cool contributions that tackle it.  Since only now it seems that there's an official solution coming, one can only remark the obvious challenge of addressing us users' requests while remaining true to the package's nature that has made it so successful all along; and thus thank the makers for keeping up with us. 

While I've thought about this recently, I'm not sure that my concerns bring something new to the table.  So I'd like to cheer up the team, and get into the package to understand the nots and bolts there.  

Cheers.

**Comment by mvabl-steve** [comment]
I've been teaching myself R from scratch, basically by undertaking something, then reading posts like these, and trial-and-error based on that. Sometimes I hit a wall and reach out.

I've hit a wall. I have dplyr 0.7 installed. I have a tibble with a column - call it `contract_key` - I added by applying mutate(coalesce()) to three other columns in the tibble. Here's sample data, which I'll of course refer to later as df:

`product,supplier,vendor,quantity,gross,contract_key
168145850127,person1,org1,7,8.14E-04,person1
655393265191,person5,org2,7,0.041861099,person5
655393265191,person3,org3,5,0.019268984,person3
168145850127,person4,org2,4,0.023245343,org2
928893912852,person5,org2,2,0.002285355,person5
NA,person6,org1,3,NA,person6
168145850127,person7,org1,1,NA,person7
350468621217,person8,org2,2,0.004212372,person8
977939797847,person9,org1,1,0.001832018,person9
168145850127,person10,org3,6,0.043907375,168145850127
168145850127,person5,org2,3,0.020960582,168145850127
168145850127,person5,org2,1,0.006923272,person5
350468621217,person11,org1,2,0.002730578,350468621217
350468621217,person5,org2,1,0.005505893,person5
977939797847,person11,org1,18,0.023260808,org1
350468621217,person12,org2,1,0.003525162,org2
168145850127,person13,org4,1,0.001932924,person13
655393265191,person5,org2,7,0.016847417,person5
NA,person14,org5,2,NA,NA
928893912852,person5,org2,2,4.22E-04,person5
168145850127,person5,org2,1,5.82E-04,person5
168145850127,person5,org2,1,0.007305813,168145850127
866586597461,person5,org2,8,0.051104768,866586597461
977939797847,person5,org2,6,0.007783279,person5
866586597461,person5,org2,4,0.0225,866586597461`

I now want to add another column that categorizes the contents of `contract_key` based on which of the three columns provided the content (through coalesce()). So if contract_key ="person5", for example, the new column, contract_level, would be "person". And contract_key="org2" would map to contract_level = "org, etc.

Here's the code I've cobbled together with my limited R:

`df <- df %>% 
  mutate(contract_key = coalesce(UPC,Artist,Label))`

I'm stumped. I've tried `if_else`, and I see that I shouldn't bother trying `case_when` (because it's inside mutate()).

It's probably basic R syntax that I just don't know. But if someone supplies the answer, I will backtrace until I figure out what you did. (And I'll have learned a new lesson in R!)

Thanks!

**Comment by friep** [comment]
@mvabl-steve I'd suggest you ask your question on https://stackoverflow.com/questions/tagged/r.  As you said yourself this is probably a question of understanding and not an issue regarding functionality. It's super valid to look for help if you're stuck but in the R community, stackoverflow is the place to go and not github issues (unless it is really an issue). Before you post there, probably read up on https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example.

**Comment by mvabl-steve** [comment]
thanks @friep. i didn't realize the unwritten rules for SO v git. (Seriously, I didn't. Sincere thanks.)

I'll try SO, and all its attendant downvoting and judging! :)
